### PR TITLE
feat(core): use literal for stack.partition (under feature flag)

### LINKED ISSUE #4092: Make `stack.partition` concrete if `region` is concrete
### Description

If users specify `env.region` when they define a `Stack`, we should be able to return a concrete value for `stack.partition` instead of `{ "Ref": "AWS::Partition" }`

### Proposed Solution

Create a map from region to partition in `region-info`.

### Environment

  - **CDK CLI Version:** 
  - **Module Version:** 
  - **OS:** 
  - **Language:**

### PR DESCRIPTION
**Pull request by jgrebholz**
closes #4092

When a Stack's region is a string literal, we can return a string
literal from Stack.partition, resulting in easier to read generated
templates.

The feature flag to enable this behavior is `@aws-cdk/core:enablePartitionLiterals`.

----

### All Submissions:

* [X] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
For any customers relying on the output of `partition()`, this will be a breaking change, as evidenced by the integration tests being broken in other parts of the code base.

I suppose that it can be submitted under a feature flag but I'm not sure the value of it is worth adding another one.

**Comment by jgrebholz** [comment]
Thanks for your review @TheRealAmazonKendra   I've updated the PR to include a feature flag, a minor update to the README, and confirmed the previously failing aws-iam integ tests are passing now.

File: packages/@aws-cdk/core/test/stack.test.ts [hunk 1]
```diff
@@ -867,6 +867,43 @@ describe('stack', () => {
     expect(stack.resolve(stack.region)).toEqual('es-norst-1');
   });
 
+  describe('@aws-cdk/core:enablePartitionLiterals', () => {
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/core/test/stack.test.ts:870]: Please make this descriptive instead of listing the flag. Better readability that way.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
Looks like there are more test failures to attend to. Note that integration tests have feature flags enabled by default so anywhere that's using this function without a token will update. This may take a few rounds to discover all the locations of the failures.

**Comment by TheRealAmazonKendra** [comment]
Oh, I almost forgot, now that this is behind a feature flag, please update your PR to reflect that. We have specific formatting instructions in our contributing guide.

File: packages/@aws-cdk/cx-api/lib/features.ts [hunk 2]
```diff
@@ -313,6 +313,29 @@ export const S3_CREATE_DEFAULT_LOGGING_POLICY = '@aws-cdk/aws-s3:createDefaultLo
...
+ * their environment, such as "aws" or "aws-cn".  Previously the CloudFormation intrinsic function
+ * "Ref: AWS::Partition" was used.  For example:
+ *
+ * Principal:
+ *   AWS:
+ *     Fn::Join:
+ *       - ""
+ *       - - "arn:"
+ *         - Ref: AWS::Partition
+ *         - :iam::123456789876:root
```
Comment (rix0rrr) [inline packages/@aws-cdk/cx-api/lib/features.ts:327]: Comments are in Markdown. This needs backticks around it to be formatted in monospace.

**Review by rix0rrr — COMMENTED** [review]
This PR changes suspiciously little. I would have expected literally every integration test to blow up, and how confident are we that RegionInfo is correct for all these regions?

File: packages/@aws-cdk/cx-api/lib/features.ts (hunk 2 above)
Comment (jgrebholz) [reply packages/@aws-cdk/cx-api/lib/features.ts:327]: Thanks for your review, I'll get this fixed in the next push

File: packages/@aws-cdk/core/test/stack.test.ts (hunk 1 above)
Comment (jgrebholz) [reply packages/@aws-cdk/core/test/stack.test.ts]: This is done in the latest update; thanks

**Comment by jgrebholz** [comment]
The most recent push includes the results of `yarn integ --update-on-failed` for each of the failing integration tests needed for my local run of `./build.sh` to succeed.  I'll review the results of the build fleet tomorrow and update as needed.  I'm not sure why so many integ tests reported UNCHANGED on their snapshot comparisons, but I'm happy to investigate any that you think should be objecting.  Is it possible they're using AWS.Region or a stack parameter for region?

The RegionInfo looks correct to me with 30 regions:
```
$ cat built-ins.generated.ts | grep FactName.PARTITION | awk '{print $3, $7;}'
"af-south-1", "aws"
"ap-east-1", "aws"
"ap-northeast-1", "aws"
"ap-northeast-2", "aws"
"ap-northeast-3", "aws"
"ap-south-1", "aws"
"ap-southeast-1", "aws"
"ap-southeast-2", "aws"
"ap-southeast-3", "aws"
"ca-central-1", "aws"
"cn-north-1", "aws-cn"
"cn-northwest-1", "aws-cn"
"eu-central-1", "aws"
"eu-north-1", "aws"
"eu-south-1", "aws"
"eu-south-2", "aws"
"eu-west-1", "aws"
"eu-west-2", "aws"
"eu-west-3", "aws"
"me-south-1", "aws"
"sa-east-1", "aws"
"us-east-1", "aws"
"us-east-2", "aws"
"us-gov-east-1", "aws-us-gov"
"us-gov-west-1", "aws-us-gov"
"us-iso-east-1", "aws-iso"
"us-iso-west-1", "aws-iso"
"us-isob-east-1", "aws-iso-b"
"us-west-1", "aws"
"us-west-2", "aws"
```

**Comment by TheRealAmazonKendra** [comment]
I'm confused about why the linter is failing. There are updates to READMEs in this change.

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**